File: CallBackInterface_exmaple/app/src/main/java/com/example/dinhtho/callbackinterface_exmaple/MainActivity.java
package com.example.dinhtho.callbackinterface_exmaple;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;


public class MainActivity extends AppCompatActivity {

    private static final String TAG = "MainActivity";
    private String s;



    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        test(100, new MyCallback() {
            @Override
            public void callbackCall(String result) {
                Log.i(TAG, "callbackCall: "+result);
            }
        });

        Log.i(TAG, "onCreate: "+s);



    }

    private void test(final int max, final MyCallback myCallback){
        Thread thread= new Thread(new Runnable() {
            @Override
            public void run() {
                int i=0;
                while (i<=max){
                    i++;
                    Log.i(TAG, "run: "+i);
                }
                s="done";
                myCallback.callbackCall("done");

            }
        });
        thread.start();
    }



    interface MyCallback {
        void callbackCall(String result);
    }
}

File: Storage/Room/app/src/main/java/com/example/learn_room/MainActivity.kt
package com.example.learn_room

import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import androidx.room.Room

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val db = Room.databaseBuilder(
            applicationContext,
            AppDatabase::class.java, "test-db"
        ).build()

        Thread({
//            db.userDao().insertAll(User("A1", "B1"))
//            db.userDao().insertAll(User("A2", "B2"))
//            db.userDao().insertAll(User("A1", "B1"))
//            db.userDao().insertAll(User("A1", "B1"))
//            db.userDao().insertAll(User("A1", "B1"))

//            db.userDao().delete(User(2, "A2", "B2"))

//            db.userDao().update("A1")
//            var users = db.userDao().getAll()
//            for (user in users) {
//                Log.d("MainActivity", "users = " + user)
//            }

//            db.bookDao().insertAll(Book("Book 1", 1))
//            db.bookDao().insertAll(Book("Book 2", 1))
//            db.bookDao().insertAll(Book("Book 3", 2))

            var books = db.bookDao().loadAllByUserId(IntArray(1){2})
            for (book in books) {
                Log.d("MainActivity", "books = " + book)
            }

        }).start()

    }
}
